extensions-contrib/materialized-view-selection/src/main/java/io/druid/query/materializedview/DerivativesManager.java

```diff
@@ -0,0 +1,246 @@
+
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Read and store derivatives information from dataSource table frequently.
+ * When optimize query, DerivativesManager offers the information about derivatives.
+ */
+@ManageLifecycleLast
+public class DerivativesManager 
+{
+  private static final EmittingLogger log = new EmittingLogger(DerivativesManager.class);
+  private final MaterializedViewConfig config;
+  private final Supplier<MetadataStorageTablesConfig> dbTables;
+  private final SQLMetadataConnector connector;
+  private final ObjectMapper objectMapper;
+  private final Object lock = new Object();
+  
+  private static volatile AtomicReference<ConcurrentHashMap<String, SortedSet<Derivative>>> derivativesRef = new AtomicReference<>(new ConcurrentHashMap<>());
+  private volatile boolean started = false;
+  private volatile ListeningScheduledExecutorService exec = null;
+  private volatile ListenableFuture<?> future = null;
+  
+  @Inject
+  public DerivativesManager(
+      MaterializedViewConfig config,
+      Supplier<MetadataStorageTablesConfig> dbTables,
+      ObjectMapper objectMapper,
+      SQLMetadataConnector connector)
+  {
+    this.config = config;
+    this.dbTables = dbTables;
+    this.objectMapper = objectMapper;
+    this.connector = connector;
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    log.info("starting derivatives manager.");
+    synchronized (lock) {
+      if (started) {
+        return;
+      }
+      exec = MoreExecutors.listeningDecorator(Execs.scheduledSingleThreaded("SQLMetadataDerivativesManager-Exec--%d"));
+      final Duration delay = config.getPollDuration().toStandardDuration();
+      future = exec.scheduleWithFixedDelay(
+          new Runnable() {
+            @Override
+            public void run() 
+            {
+              try {
+                updateDerivatives();
+              }
+              catch (Exception e) {
+                log.makeAlert(e, "uncaught exception in derivatives manager updating thread").emit();
+              }
+            }
+          },
+          0,
+          delay.getMillis(),
+          TimeUnit.MILLISECONDS
+      );
+      started = true;
+    }
+    log.info("Derivatives manager started.");
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (lock) {
+      if (!started) {
+        return;
+      }
+      started = false;
+      future.cancel(true);
+      future = null;
+      final ConcurrentHashMap<String, SortedSet<Derivative>> emptyMap = new ConcurrentHashMap<>();
+      ConcurrentHashMap<String, SortedSet<Derivative>> current;
+      do {
+        current = derivativesRef.get();
+      } while (!derivativesRef.compareAndSet(current, emptyMap));
```

Comment:
Would you elaborate more on why clearing  the map inside `derivativesRef` is needed?